risc0/zkvm/platform/io.h

```diff
@@ -16,54 +16,114 @@
+struct ShaDescriptor {
+  // Number of 64-byte chunks of SHA data to process
+  uint32_t typeAndCount;
 
-struct GetKeyDescriptor {
-  uint32_t name;
-  uint32_t addr;
-  uint32_t mode;
+  // TODO(nils): "idx" seems to be always zero; is it used for anything?
+  uint32_t idx;
+
+  // Address of first byte of SHA data to process
+  uint32_t source;
+
+  // Output address to write SHA digest to; this memory must be uninitialized, and must not be
+  // modified afterwards.
+  uint32_t digest;
 };
 
 inline volatile ShaDescriptor* volatile* GPIO_SHA() {
   return reinterpret_cast<volatile ShaDescriptor* volatile*>(kGPIO_SHA);
 }
 
-inline volatile IoDescriptor* volatile* GPIO_Write() {
-  return reinterpret_cast<volatile IoDescriptor* volatile*>(kGPIO_Write);
-}
+// To commit to output data, the guest fills in IoDescriptor structure
+// with the size and address of the buffer to write, and then writes
+// its address to kGPIO_Commit.
+
+struct IoDescriptor {
+  uint32_t size;
+  uint32_t addr;
+};
 
 inline volatile IoDescriptor* volatile* GPIO_Commit() {
   return reinterpret_cast<volatile IoDescriptor* volatile*>(kGPIO_Commit);
 }
 
-inline volatile FaultDescriptor* volatile* GPIO_Fault() {
-  return reinterpret_cast<volatile FaultDescriptor* volatile*>(kGPIO_Fault);
+// To execute general IO, the guest first ensures GPIO_SendRecvChannel
+// is filled with the application-defined channel number and GPIO_SendRecvSize
+// with the number of bytes in the message.
+//
+// The guest then writes the address of the first byte of the message to
+// sned to GPIO_SendRecvAddr; when this write is performed, the host
+// passes the given message to the host application layer.
+//
+// Once the application layer has completed processing the given message,
+// it supplies a sequence of bytes as a reply (which may be length 0).
+//
+// The host will then write this result into the guest's Input memory
+// region as a word indicating the numbre of bytes in the reply,
+// followed by the contents of the reply padded to the next word.
+//
+inline uint32_t volatile* GPIO_SendRecvChannel() {
+  return reinterpret_cast<uint32_t volatile*>(kGPIO_SendRecvChannel);
+}
+
+inline uint32_t volatile* GPIO_SendRecvSize() {
+  return reinterpret_cast<uint32_t volatile*>(kGPIO_SendRecvSize);
 }
 
-inline volatile LogDescriptor* volatile* GPIO_Log() {
-  return reinterpret_cast<volatile LogDescriptor* volatile*>(kGPIO_Log);
+inline const void* volatile* GPIO_SendRecvAddr() {
+  return reinterpret_cast<const void* volatile*>(kGPIO_SendRecvAddr);
 }
 
+// To raise a fatal error, the guest writes the address of a
+// null-terminated string describing the fatal error to GPIO_Fault.
+inline const char* volatile* GPIO_Fault() {
+  return reinterpret_cast<const char* volatile*>(kGPIO_Fault);
+}
+
+// To log an informative message, the guest writes the address of
+// a null-terminated string containing the log message to GPIO_Log.
+inline const char* volatile* GPIO_Log() {
```

Comment:
I don't think this is plumbed up to Rust (or if it is I didn't see if in the guest crate).  Had a case on Discord where I was considering suggesting this...